src/main/java/org/opensearch/ad/task/ADTaskManager.java

```diff
@@ -2968,4 +2878,111 @@ public int getLocalAdAssignedBatchTaskSlot() {
         return adTaskCacheManager.getTotalDetectorTaskSlots();
     }
 
+    // =========================================================
+    // Methods below are maintenance code triggered by hourly cron
+    // =========================================================
+
+    /**
+     * Maintain running historical tasks.
+     * Search current running latest tasks, then maintain tasks one by one.
+     * Get task profile to check if task is really running on worker node.
+     * 1. If not running, reset task state as STOPPED.
+     * 2. If task is running and task for HC detector, check if there is any stale running entities and
+     *    clean up.
+     *
+     * @param transportService transport service
+     * @param size return how many tasks
+     */
+    public void maintainRunningHistoricalTasks(TransportService transportService, int size) {
```

Comment:
Do you think it is necessary to add maintenance for memory tracker just in case you don't release memory? I have one in PriorityCache for HC detector: https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/caching/PriorityCache.java#L517-L539

Reply:
For historical analysis, we will clean up cache if task finished normally or exception happens. If we fail to remove cache, the reset logic will try to clean up again, and hourly cron will clean up as well. So I think it's safe. But I'm open to add more check to make sure no memory leak. Will create an issue to track this.

Reply:
Sounds good to me.